design/003-service-topology-support.md

```diff
@@ -0,0 +1,118 @@
+
+Toplogy metadata that is obtained from supporting this feature allows for weighting endpoints based on locality.
+
+(_Adapted from [#4325](https://github.com/linkerd/linkerd2/issues/4325)_)
+(_N.B: I use locality and topology interchangeably here, when used in the same context to me both mean the same thing, apologies in advance if the language barrier created any confusion._)
+
+## Design proposal (Step 2)
+
+[design-proposal]: #design-proposal
+
+### Summary
+---
+
+To implement support for this feature, my main assumption is that most of the work will be carried out in the `destination service` on the control plane side. Although the feature does not add a lot of changes in terms of the Kubernetes API, several changes have to be made. The first and most obvious to me is supporting the `endpointSlices` feature that was introduced in Kubernetes `v1.16`, since this will enable endpoints to hold topology metadata.
+
+All meshed pods have a proxy that will first establish all outbound endpoints of a service the pod is targeting. Through this change, we will introduce a new weight to each endpoint based on locality, which can afterwards be used to add subsequent features that build on locality-awareness. I propose that the weights be added to each endpoint on the proxy side. To ensure backwards compatibility, there should be a separate gRPC call that can be made to the destination service to get a service's topology preference, and the host's (node) topology labels. The call would include the FQDN/IP of a service, and a topology identifier (similar to the context namespace used for service profiles). 
+
+The destination service will resolve the FQDN/IP in its usual fashion; it will now also hold in cache the endpoint slices of the service (which have topology metadata). As part of every service that is cached, the destination service should add (if they exist) the topology keys in the order of preference of the owner. Based on the locality identifier of the source, the topology preference of the service, and the available endpoints, each endpoint can have a topology weight assigned to it. The proxy receives its endpoint(s) as usual and can assign weights based on the information it receives from the separate topology call and the topology metadata associated with each endpoint(s).
```

Comment:
Do you have a handle on how `EndpointSlice` will impact our discovery? What does the watch look like for the proxy and destination service?

Reply:
From what I currently understand, this is the whole process (including the watch) for service discovery:

Based on the FQDN/IP of the service the pod is trying to access, the proxy will build a tower stack for the outbound traffic. The stack includes the load balancing layer which wraps around a discovery crate and a resolver. A connection is established to the destination service via the discovery layer; the layer contains a buffer and a watchdog to poll for updates. (Apologies if so far any of this is wrong, still getting up to grips with Rust).

Meanwhile, in the destination service the server accepts calls from the proxy in the form of `Destination.Get`. The service name is at first resolved, a set of weighted address set is sent back to the proxy. Every time a new address is discovered for a service, a new weighted set is created with its endpoint and sent to the proxy. All CRUD type updates are propagated from the destination service to the proxy.

The destination service actively watches and caches all services and endpoints (all except whatever is in kube-system ns). Each service keeps track of its ports by identifying them through their hostname and port number.

Endpoint slices differ from endpoints through a few spec changes:
1. Hostnames are no longer included as part of an address entry, instead this has been replaced by the topology label map;
2. Endpoints contain ALL network endpoints in a single resource, endpoint slices contain by default 100 addresses. Because of this, there is no longer a 1:1 relation between an endpoints resource and a service. Instead, a service can have many endpoint slices, so each endpoint slice name (id) is randomly generated based on the service name, the service it belongs to has to be inferred from its labels.
3. Endpoint slices support dual-stack endpoints, so an address entry can, in theory, contain both an IPv4 and an IPv6.

If I understand the question correctly, by introducing `EndpointSlice` most of the work on the discovery watcher will be geared around making it work with current implementation of the port/service publishers, beyond that at the moment I can't see anything else that will have to be changed, but in case I have it wrong I can keep on looking and explore it a bit more in-depth.

I haven't explored much (yet) how the `portPublisher` will be affected by these changes, but my assumption is that the subscription methods will not be affected by these changes, and even though addresses are split between many resources it is still possible to consolidate them into a single `portPublisher` when they are mapped to the same port.

tl;dr at the moment I think the impact will be mainly on integrating the `EndpointSlices` discovery informer and its handler functions to work with all of the code in the `endpoint_watcher`. Instead of a 1:1 mapping between svcs and endpoints we (can) now have many endpoint slices resources mapped to the same service so they would all have to be kept in sync and consolidated.